### ISSUE TITLE
Animated item textures in the inventory

### ISSUE THREAD

**Issue by bigfoot547** [luanti-org/luanti#4758]
I think a good addition to the game would be animatd textures in the inventory.

**Comment by paramat** [comment]
Do you know we have an option for spinning inventory items when wielded?
Anything more would be distracting to the eye and completely unnecessary.

**Comment by bigfoot547** [comment]
Wow! How?

**Comment by sofar** [comment]
change this setting in `minetest.conf`:
```
inventory_items_animations = true
```

**Comment by kaeza** [comment]
I think he means animated textures, not just rotating items.

**Comment by paramat** [comment]
@bigfoot547 could you clarify what you meant? Do the spinning items satisfy your request?

**Comment by sofar** [comment]
He specifically states "animatd textures" (sic).

**Comment by numberZero** [comment]
@paramat Animation may be cool, if not overused. Just some sparking of diamonds, for example.
It could be activated by mouse hover and sometimes randomly (if applicable for the item; e.g. a diamond can spark by itself, but a pick should remain still). (Something like that is implemented for creatures on the battlefield in HoMM3 and HoMM3 HotA, for example).

Currently, blocks are animated rather well, although could spin a bit faster. For manual tools like picks, spinning is acceptable although too slow, that slowness annoys (instant stop annoys too, but probably wouldn’t with faster spinning).

Also, spinning of a 2d image often looks just weird. Have you ever seen spinning 2D cart?

P.S. And no, no animation in the hotbar! That’s *really* distracting!

**Comment by C1ffisme** [comment]
@numberZero 
> P.S. And no, no animation in the hotbar! That’s really distracting!

Agreed, it just looks silly when a tool is rotating.

**Comment by paramat** [comment]
I understand this as meaning that nodes with animated textures (water, lava) appear in the inventory as they do in the game.

**Comment by kaeza** [comment]
I understood this as a sort of animated `inventory_image` (e.g. to animate the fire on torches or something).

**Comment by paramat** [comment]
That would then need another whole set of animated frames just for the inventory, that would be excessively fancy and most texture makers could not be bothered.

**Comment by numberZero** [comment]
@paramat They might create an overlay to use on many items. E.g. general `ingotlike` sparking, `diamondlike`, etc.
In many cases it could be nice to just rotate meshes, though (the new MT style seems to be full of meshes anyway). Or define other animation styles.
I understand that this is not so trivial to do well, but the result would be impressive IMO.

**Comment by bigfoot547** [comment]
Sorry for the late responce.
I intended this issue to suggest an idea to have craftitems have animations also.
Yes, animations should be in the hotbar, hand and regular inventory. Maybe there could be a setting.
Craftitem and tools animations should be shown **JUST LIKE NODES** in the way that it is run **CONTINUOSLY** and left up to the modders descretion. The players can give feedback, and the modder can take it or leave it.(Mods are on the forums for a reason!) If someone doesn't like it, they can rewrite the mod, and not have as many animations.
That is my piece, take it or leave it.

    --bigfoot547

**Comment by kaeza** [comment]
#4297 ?

**Comment by PureTryOut** [comment]
I'd like this feature as well. I'm not entirely sure why this is hard to understand for some people, and saying "it would be distracting" is not an argument as that's entirely subjective. Honestly, there could probably be a client-side setting to disable such animations either entirely, only in hotbar, only in hand, etc.

Rotating is *not* enough. We might like a way to:
* animate the flame of a torch so it flickers when holding it
* make an enchantment mod which has a sparkling "glow" over the items texture
* make a clock or compass
* make an in-game map of some sorts

The possibilities are endless really.

**Comment by C1ffisme** [comment]
@paramat

> That would then need another whole set of animated frames just for the inventory, that would be excessively fancy and most texture makers could not be bothered.

What?

> most texture makers could not be bothered.

What?!?

> that would be excessively fancy

What?!?!

I'm sorry, but is every single feature that may or may not cause lag "excessively fancy"? This is a **game engine**. You can do what you want with the API provided, _but only if the API supports that feature_, or a hacky solution is found.

As for most texture designers being too lazy to animate items, this simply isn't true. Just look at Minecraft's Resource pack collection. People have bothered to change the game's textures, sounds, text, and _even the models of blocks_!

Besides, if you want to make a texture pack that looks anywhere close to good, you don't just spend 5 minutes copying or pasting textures. For heaven's sake, we've debated the grass texture and changed it at least four times!

**Comment by paramat** [comment]
> I'm sorry, but is every single feature that may cause the tiniest bit of lag "excessively fancy"?

'Excessively fancy' is just my opinion of the feature and has nothing to do with the more intensive code.

>  This is a game engine. You can do what you want with the API provided, but only if the API supports that feature,

The game engine argument cannot be used to justify adding every feature imaginable.

My complaint about distraction was silly. I'm more neutral now.

**Comment by Avamander** [comment]
Well to be honest this is not just "some feature", it's something that's rather basic in terms of functionality if you compare Minetest to many other games, a lot of engines do have animated inventory or toolbar items.

**Comment by bigfoot547** [comment]
The API would be like the tile animation definition.

**Comment by C1ffisme** [comment]
> The game engine argument cannot be used to justify adding every feature imaginable.

No, but it's worth trying to cover most, if not all of the needs that developers for sandbox games will require.

**Comment by paramat** [comment]
> it's something that's rather basic in terms of functionality if you compare Minetest to many other engines

MT is not like other games, it's deliberately simple. It's not a basic or essential feature as we're fine (and have been fine) without it.

**Comment by Avamander** [comment]
@paramat You are confusing simple and feature-lacking.

**Comment by bigfoot547** [comment]
If this game were purely "simple", we wouldn't be in 0.4.15-dev. We would be in 1.0.0-stable.
This game is like minecraft indev, infdev, survival test, ans pre-classic in 1. It is not a complete game. Minecraft is a complete game because they only add features that are just "fluff". We need to add *real* features and game mechanics instead of "fluff".

**Comment by bigfoot547** [comment]
This would be a game mechanic.

**Comment by paramat** [comment]
Any thoughts about MT 'completedness' are nonsensical, it's not black and white, MT will never be complete but also has been complete for years. MT will always be lacking in features because new features can always be added.
I'm not opposed to the feature, it's just not a high priority.

**Comment by C1ffisme** [comment]
> Any thoughts about MT 'completedness' are nonsensical, it's not black and white, MT will never be complete but also has been complete for years.

Either it's complete, it's not, or we have no idea what we're doing.

* If the game is complete, then the game should not be shipped with the current 865 issues that it has (give or take a few feature requests like this one), or it would be stable. The only versions with issues should be development versions.

* If the game is not complete, then we should keep working towards our goal, towards the goal of a finished, and (mostly) perfect product. The problem is that we all have different visions for the game, and the goal for the final product is still in debate. Which brings us to our third option:

* If we have no idea what we are doing, we have no goal in sight, and we are arguing pointlessly. It's fine to not have a clear goal at first, but an idea of what belongs and what doesn't is at least helpful, and gets us out of a state of unorganized, messy development, as well as saving many of us a lot of valuable time that can be spent coding features.

I'm okay with not being sure what this game will turn out to be. No one really knows that. But if we can't even decide whether or not it has already reached that point, **we're in huge trouble**.

**Comment by sofar** [comment]
The engine should make this possible, the subgame should choose to use the feature (or not).

I don't think this would affect performance much. Animating textures isn't that expensive, but keep in mind that the way that the inventory items are drawn doesn't offer any compatible way to animate textures, really (only for nodes, but not for items or tools).

I wouldn't dismiss any PR that would properly implement this, though, on the contrary.

**Comment by paramat** [comment]
> Either it's complete, it's not, or we have no idea what we're doing.

A game is never complete because work always conitnues, there are always new features to add, improvements, bugfixes, optimisations. A point may come where we are very satisfied, but then a little later we will change our mind and realise 'actually, we really need this', so deciding 'completedness' is impossible.
There will always be issues even if 'complete'.

> we should keep working towards our goal, towards the goal of a finished, and (mostly) perfect product

This is ridiculous simplistic thinking, there is no need for a static 'goal', just an intention for continual improvement, which is our intention already.

In another way MT has been complete, as in very usable and playable, for years.
We obviously know what we're doing, maybe you don't but that's your problem.

**Comment by C1ffisme** [comment]
> A game is never complete because work always conitnues

That was not always true, and isn't always true today. It used to be that a game would stabilize itself, and then be sold on a disc. It was only when the internet rolled around that we could fix bugs and add new features to the game.

> so deciding 'completedness' is impossible

Not entirely. It depends on your definition of "complete". Minecraft reached a "Complete" stage, but still kept adding features. We don't have to stop adding features, I would just like to see a "stable 1.0.0" version that is:
* Actually Fun (First priority of a game)
* Contains things that would be expected of a game of this type.
* Very few or no bugs, not including ones made by using 3rd party software.

> We obviously know what we're doing, maybe you don't but that's your problem.

It doesn't feel like that at times. Sometimes it feels like when one person wants to drag the game in the "fun, detailed and graphical pleasing" direction, but someone else is pulling against it towards the "simplistic, almost lazy graphics and fun for people with high attention spans" area.

**Comment by PureTryOut** [comment]
Could we get back on-topic please?

**Comment by C1ffisme** [comment]
@PureTryOut Yeah, I can agree with that. @paramat You can have the last word if you want to, I'm getting kinda tired of arguing.

**Comment by sofar** [comment]
Note that this issue isn't gonna go anywhere soon:

1) active devs are busy with things much more urgent
2) most people commenting on this thread lack the knowledge to code this

This is a trend in issues and is sadly not very useful. There's no need to push back against @paramat because if someone codes it up nicely and it passes review, I'm sure that it will be considered fairly (and thus merged if technically in order).

Perhaps though, features like this should go into a nice big visual enhancement forum thread instead.

**Comment by LoneWolfHT** [comment]
*bump*
I'd like to be able to animate my item textures instead of using a separate item for each frame

**Comment by loffren174** [comment]
Has anything in the overall situation changed regarding this issue?
Having animated inventory images is a desirable feature on its own to some people (including me),
but it would also have an additional use: 

You could finally play animated images on the HUD, by simply showing an inventory slot on the hud with the item within it playing the desired inventory-image-animation.

Usage samples:
radio-symbol of outgoing radio-communication waves next to some incoming message (alert, story dialog, etc)
any occasional glint or actually animated HUD element whatsoever (battery energy bar glinting occasionally, stamina bar/icon gently flashing when empty and player attempts to use it,...)

There's animated image support for formspecs (even support for 9-sliced ones coming up in #10265 ), but I feel that waiting for direct and complete support of formspec elements in HUD has much less chance of happening in the coming year(s). There's a formspec-to-hud mod in the content-db, but it doesn't support the animated images either, so this feels the best chance to get there to me.

The opposite approach would also work (only supporting animated HUD images, and using those to draw 'over' an inventory slot to emulate animated inventory items), but this seems a more straightforward method. with a direct usecase additional to the one I personally would like.

**Comment by DeletedUser56** [comment]
> Do you know we have an option for spinning inventory items when wielded?
> Anything more would be ~distracting to the eye and~ completely unnecessary.

![mcl_enchanting_book_enchanted](https://user-images.githubusercontent.com/81821021/125195539-11302500-e224-11eb-99fe-43eeadb5bd71.png)
why unnessisary? I think it would be a great feature.

**Comment by sfan5** [comment]
This should be rather easy to implement and even the interaction with `inventory_items_animations` isn't that problematic.

**Comment by farooqkz** [comment]
Today I saw these:
https://mastodon.gamedev.place/@birdibirdson/113216211302602125

And I thought CTF could sure make good use of them. Then I realized there is no support for them in Minetest and found this issue.

**Comment by ezileicorreia** [comment]
> alterar esta configuração em `minetest.conf`:
> 
> ```
> inventory_items_animations = true
> ```

I was looking for this, I thought it was a mod.
